File: 3781-maximize-the-distance-between-points-on-a-square/maximize-the-distance-between-points-on-a-square.java
import java.util.Arrays;

class Solution {
    // Unroll the 2D perimeter into a 1D line segment [0, 4*side)
    private long mapTo1D(int x, int y, int s) {
        if (y == 0) return (long) x;                             // Bottom edge
        if (x == s) return (long) s + y;                         // Right edge
        if (y == s) return 2L * s + (s - x);                     // Top edge
        if (x == 0) return 3L * s + (s - y);                     // Left edge
        return 0;
    }

    private boolean check(long mid, int N, int k, long[] A2, long totalLen) {
        int n2 = 2 * N;
        int[] nxt = new int[n2 + 1];
        nxt[n2] = n2; 
        
        // Two-pointer precomputation of the next valid point
        int j = 0;
        for (int i = 0; i < n2; ++i) {
            while (j < n2 && A2[j] - A2[i] < mid) {
                j++;
            }
            nxt[i] = j;
        }

        // Try placing K points starting from every possible initial point
        for (int i = 0; i < N; ++i) {
            int curr = i;
            int steps = 1;
            while (steps < k && curr < n2) {
                curr = nxt[curr];
                steps++;
            }
            // Check if the k-th point successfully wrapped around without overlapping
            if (curr < n2 && A2[curr] <= A2[i] + totalLen - mid) {
                return true;
            }
        }
        return false;
    }

    public int maxDistance(int side, int[][] points, int k) {
        int N = points.length;
        long[] A = new long[N];
        
        for (int i = 0; i < N; ++i) {
            A[i] = mapTo1D(points[i][0], points[i][1], side);
        }
        Arrays.sort(A);

        // Duplicate the array to handle circular nature
        long[] A2 = new long[2 * N];
        long totalLen = 4L * side;
        for (int i = 0; i < N; ++i) {
            A2[i] = A[i];
            A2[i + N] = A[i] + totalLen;
        }

        // Binary search for the maximum possible minimum distance
        long low = 1, high = side;
        long ans = 0;

        while (low <= high) {
            long mid = low + (high - low) / 2;
            if (check(mid, N, k, A2, totalLen)) {
                ans = mid;
                low = mid + 1; // Try for a larger minimum distance
            } else {
                high = mid - 1; // Distance too large, scale back
            }
        }

        return (int) ans;
    }
}

File: 3781-maximize-the-distance-between-points-on-a-square/maximize-the-distance-between-points-on-a-square.cpp
class Solution {
public:
    using ll = long long;

    int maxDistance(int side, vector<vector<int>>& points, int k) {
        int n = points.size();
        ll L = 4LL * side;

        vector<ll> pos;

        for (auto &p : points) {
            ll x = p[0], y = p[1];

            if (y == 0) pos.push_back(x);
            else if (x == side) pos.push_back(side + y);
            else if (y == side) pos.push_back(3LL * side - x);
            else pos.push_back(4LL * side - y);
        }

        sort(pos.begin(), pos.end());

        auto check = [&](ll d) -> bool {
            vector<ll> ext = pos;
            for (ll x : pos) ext.push_back(x + L);

            int m = ext.size();

            vector<int> nxt(m);

            int j = 0;
            for (int i = 0; i < m; i++) {
                if (j < i + 1) j = i + 1;

                while (j < m && ext[j] - ext[i] < d) j++;

                nxt[i] = j;
            }

            const int LOG = 6; // k <= 25

            vector<vector<int>> up(LOG, vector<int>(m, m));

            for (int i = 0; i < m; i++)
                up[0][i] = min(nxt[i], m);

            for (int b = 1; b < LOG; b++) {
                for (int i = 0; i < m; i++) {
                    int mid = up[b - 1][i];
                    up[b][i] = (mid < m ? up[b - 1][mid] : m);
                }
            }

            for (int start = 0; start < n; start++) {
                int cur = start;
                int need = k - 1;

                for (int b = 0; b < LOG; b++) {
                    if (need & (1 << b)) {
                        cur = up[b][cur];
                        if (cur >= start + n) break;
                    }
                }

                if (cur >= start + n) continue;

                ll gap = ext[start + n] - ext[cur];

                if (gap >= d) return true;
            }

            return false;
        };

        ll lo = 0, hi = 2LL * side;

        while (lo < hi) {
            ll mid = (lo + hi + 1) >> 1;

            if (check(mid))
                lo = mid;
            else
                hi = mid - 1;
        }

        return (int)lo;
    }
};